lib/AST/ParameterPack.cpp

```diff
@@ -0,0 +1,281 @@
+
+#include "swift/AST/Types.h"
+#include "swift/AST/Decl.h"
+#include "swift/AST/ParameterList.h"
+#include "swift/AST/Type.h"
+#include "llvm/ADT/SmallVector.h"
+
+using namespace swift;
+
+void TypeBase::getTypeParameterPacks(
+    SmallVectorImpl<Type> &rootParameterPacks) const {
+  llvm::SmallDenseSet<CanType, 2> visited;
+
+  auto recordType = [&](Type t) {
+    if (visited.insert(t->getCanonicalType()).second)
+      rootParameterPacks.push_back(t);
+  };
+
+  Type(const_cast<TypeBase *>(this)).visit([&](Type t) {
+    if (auto *paramTy = t->getAs<GenericTypeParamType>()) {
+      if (paramTy->isParameterPack()) {
+        recordType(paramTy);
+      }
+    } else if (auto *archetypeTy = t->getAs<PackArchetypeType>()) {
+      if (archetypeTy->isRoot()) {
+        recordType(t);
+      }
+    }
+  });
+}
+
+bool GenericTypeParamType::isParameterPack() const {
+  if (auto param = getDecl()) {
+    return param->isParameterPack();
+  }
+
+  auto fixedNum = ParamOrDepthIndex.get<DepthIndexTy>();
+  return (fixedNum & GenericTypeParamType::TYPE_SEQUENCE_BIT) ==
+         GenericTypeParamType::TYPE_SEQUENCE_BIT;
+}
+
+/// G<{X1, ..., Xn}, {Y1, ..., Yn}>... => {G<X1, Y1>, ..., G<Xn, Yn>}...
+PackExpansionType *PackExpansionType::expand() {
+  auto countType = getCountType();
+  auto *countPack = countType->getAs<PackType>();
+  if (countPack == nullptr)
+    return this;
+
+  auto patternType = getPatternType();
+  if (patternType->is<PackType>())
+    return this;
+
+  unsigned j = 0;
+  SmallVector<Type, 4> expandedTypes;
+  for (auto type : countPack->getElementTypes()) {
+    Type expandedCount;
+    if (auto *expansion = type->getAs<PackExpansionType>())
+      expandedCount = expansion->getCountType();
+
+    auto expandedPattern = patternType.transformRec(
+      [&](Type t) -> Optional<Type> {
+        if (t->is<PackExpansionType>())
+          return t;
+
+        if (auto *nestedPack = t->getAs<PackType>()) {
+          auto nestedPackElts = nestedPack->getElementTypes();
+          if (j < nestedPackElts.size()) {
+            if (expandedCount) {
+              if (auto *expansion = nestedPackElts[j]->getAs<PackExpansionType>())
+                return expansion->getPatternType();
+            } else {
+              return nestedPackElts[j];
+            }
+          }
+
+          return ErrorType::get(t->getASTContext());
+        }
+
+        return None;
+      });
+
+    if (expandedCount) {
+      expandedTypes.push_back(PackExpansionType::get(expandedPattern,
+                                                     expandedCount));
+    } else {
+      expandedTypes.push_back(expandedPattern);
+    }
+
+    ++j;
+  }
+
+  auto *packType = PackType::get(getASTContext(), expandedTypes);
+  return PackExpansionType::get(packType, countType);
+}
+
+bool TupleType::containsPackExpansionType() const {
+  for (auto elt : getElements()) {
+    if (elt.getType()->is<PackExpansionType>())
```

Comment:
Why not make this a property of a tuple type?

Reply:
That's a good idea. I wanted to wait to micro-optimize until more code is fleshed out though.